include/sys/util.h

```diff
@@ -426,6 +426,30 @@ char *utf8_lcpy(char *dst, const char *src, size_t n);
 /** @brief Number of Hz in @p x MHz */
 #define MHZ(x) (KHZ(x) * 1000)
 
+/**
+ * @brief Wait for an expression to return true with a timeout
+ *
+ * Spin on an expression with a timeout and optional delay between iterations
+ *
+ * Commonly needed when waiting on hardware to complete an asynchronous
+ * request to read/write/initialize/reset, but useful for any expression.
+ *
+ * @param expr Trueth expression you wish to spin on, ex: XYZREG & XYZREG_EN
+ * @param timeout Timeout to wait for in microseconds, 1000 (1ms)
+ * @param delay_stmt Delay statement, ex: NULL, k_yield(), k_msleep(1) or k_busy_wait(1)
+ *
+ * @retval expr As a boolean return, if false then it has timed out.
+ */
+#define wait_for(expr, timeout, delay_stmt)                                                        \
+	({                                                                                         \
```

Comment:
A note of surprise on my part.  I would have thought that this would have been encompassed within a do {} while (0) as opposed to ({....})

Reply:
This lets me effectively return the (expr) as a boolean from the macro, doing that with do { } while(0) could certainly be done as well

Reply:
Aah. Now I understand why that was done.

However, this leads to an observation of a possible race condition. That is, the final (expr) may be different from the (expr) evaluated in the while loop.

Perhaps we should be caching the result of (expr) in the while loop, and use that as the "return value".